src/goto-instrument/code_contracts.cpp

```diff
@@ -282,9 +354,345 @@ const symbolt &code_contractst::new_tmp_symbol(
+    exprt op_addr =
+      exprt(ID_address_of, pointer_type(curr_op.type()), {curr_op});
+
+    // Declare a new symbol to stand in for the reference
+    symbol_exprt standin =
+      new_tmp_symbol(
+        pointer_type(curr_op.type()), f_sym.location, func_id, f_sym.mode)
+        .symbol_expr();
+
+    created_decls.add(goto_programt::make_decl(standin, f_sym.location));
+
+    created_decls.add(goto_programt::make_assignment(
+      code_assignt(standin, std::move(op_addr)), f_sym.location));
+
+    // Add a map entry from the original operand to the new symbol
+    created_references.push_back(standin);
+  }
+}
+
+void code_contractst::instrument_assn_statement(
+  goto_programt::instructionst::iterator &ins_it,
+  goto_programt &program,
+  exprt &assigns,
+  std::vector<exprt> &assigns_references,
+  std::set<exprt> &freely_assignable_exprs)
+{
+  INVARIANT(
+    ins_it->is_assign(),
+    "The first argument of instrument_assn_statement should always be"
+    " an assignment");
+  const exprt &lhs = ins_it->get_assign().lhs();
+  if(freely_assignable_exprs.find(lhs) != freely_assignable_exprs.end())
+  {
+    return;
+  }
+  exprt alias_expr = create_alias_expression(assigns, lhs, assigns_references);
+
+  goto_programt alias_assertion;
+  alias_assertion.add(
+    goto_programt::make_assertion(alias_expr, ins_it->source_location));
+  program.insert_before_swap(ins_it, alias_assertion);
+  std::advance(ins_it, 1);
+}
+
+void code_contractst::instrument_call_statement(
+  goto_programt::instructionst::iterator &ins_it,
+  goto_programt &program,
+  exprt &assigns,
+  std::vector<exprt> &aliasable_references,
+  std::set<exprt> &freely_assignable_exprs)
+{
+  INVARIANT(
+    ins_it->is_function_call(),
+    "The first argument of instrument_call_statement should always be "
+    "a function call");
+  code_function_callt call = ins_it->get_function_call();
+  const irep_idt &called_name =
+    to_symbol_expr(call.function()).get_identifier();
+
+  if(std::strcmp(called_name.c_str(), "malloc") == 0)
+  {
+    aliasable_references.push_back(call.lhs());
+
+    // Make the variable, where result of malloc is stored, freely assignable.
+    goto_programt::instructionst::iterator local_ins_it = ins_it;
+    local_ins_it++;
+    if(
+      local_ins_it->is_assign() &&
+      local_ins_it->get_assign().lhs().is_not_nil())
+    {
+      freely_assignable_exprs.insert(local_ins_it->get_assign().lhs());
+    }
+    return; // assume malloc edits no currently-existing memory objects.
+  }
+
+  if(call.lhs().is_not_nil())
+  {
+    exprt alias_expr =
+      create_alias_expression(assigns, call.lhs(), aliasable_references);
+
+    goto_programt alias_assertion;
+    alias_assertion.add(
+      goto_programt::make_assertion(alias_expr, ins_it->source_location));
+    program.insert_before_swap(ins_it, alias_assertion);
+    ++ins_it;
+  }
+
+  // TODO we don't handle function pointers
```

Comment:
This is a good point, though I don't think it should be a blocker for this PR. Should I create a github issue to track this request? @hannes-steffenhagen-diffblue @tautschnig @martin-cs

Reply:
It is fine to simply require that this is run after function pointer removal.  ( Yes, there should be a system / type annotation on the goto-functions to denote / check / enforce this. ) (Yes, it would be good to get more things working before function pointer removal, I am trying.)

Agreed that it is not a blocker.  Making sure that it is run after function pointer removal and documenting it is fine.  If you want to go above and beyond, a test case would be great! :-)

Reply:
I guess the simplest change would be to make this an `INVARIANT`.